Add function to open Xcodeproj

cipolleschi opened the pull request:
Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that runs openx the xcodeproject once the integration is done.

## Changelog:
[Internal] -

Differential Revision: D81778466

cortinico reviewed (approved):
Review automatically exported from Phabricator review in Meta.

cipolleschi committed:
Add simple function to handle symlinks

Summary:
This change adds a simple function that manages symlinks. This is used through the stack of changes to avoid code repetition.

## Context
SwiftPM is very picky in how the header structure must be.
In order to preserve the import/include statements as much as possible when building from source, we can recreate the header structure in a temporary folder inside the react-native package using symlinks.

In this way, users can still modify the headers and build RNTester and HelloWorld using SwiftPM without breaking changes.

## Changelog:
[Internal] -

Differential Revision: D82202860

Reviewed By: cortinico

cipolleschi committed:
Add function to extract headers from a folder

Summary:
This change add a utility function to extract headers (both .h and .hpp) from a folder. It also allow to exclude some specific folders (e.g.: `tests`, non supported platforms, ...)

## Changelog:
[Internal] -

Differential Revision: D82205692

Commit:
Create script to create hard links for headers from a source path to a destination path (#53620)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53620

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links between a source folder and a destination folder.

## Changelog:
[Internal] -

Differential Revision: D81778469

Reviewed By: cortinico

Commit:
Create utility function to create the right folder structure for headers in the ReactApple folder (#53629)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53629

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding a function to properly create the header structure for the headers in ReactApple that requires a special case.

## Changelog:
[Internal] -

Differential Revision: D81778471

Reviewed By: cortinico

Commit:
add function hard links for headers from ReactCodegen

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function to create hard links to Codegen Headers so that other libraries can access them.

## Changelog:
[Internal] -

Differential Revision: D81778450

Commit:
Create utility function to create the right folder structure for headers in the ReactCommon folder (#53631)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53631

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links for the ReactCommon folder that has a complex structure and complex mapping requirements

## Changelog:
[Internal] -

Differential Revision: D81778470

Reviewed By: cortinico

Commit:
Add function to create hard links for headers of the third-party dependencies

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links for the ThirdPartyDependencies that React native and libraries might need to access.

## Changelog:
[Internal] -

Differential Revision: D81778459

Commit:
Create the orchestration function to coordinate the creation of the hard links

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding a function that is used to orchestrate the helpers function defined in previous changes.

## Changelog:
[Internal] -

Differential Revision: D81778457

Commit:
Create function to create hardlink to header required to build React Native Core

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we did two things:
1. we move the helper functions to a headers-utils.js file (and the tests to the `headers-utils-test.js` file)
2. we added a function that coordinates the creation of the header links *and added tests for it)

## Changelog:
[Internal] -

Differential Revision: D81778464

Commit:
Add function to generate artifact for codegen (#53676)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53676

## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that generates codegen artifacts. Those are required by React Native and by the app to build properly.

## Changelog:
[Internal] -

Differential Revision: D81778439

Commit:
Add function to integrate Swift PM into the Xcodeproject

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that integrates Swift packages into the xcodeproject

## Changelog:
[Internal] -

Differential Revision: D81778438

Commit:
Add script to add the RN Swift packages to xcodeproject

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds an higeher level function that can be used to integrate local swift packages into Xcodeproject files.

This will also be helpful for autolinking packages.

## Changelog:
[Internal] -

Differential Revision: D81778453

Commit:
Add function to coordinate the headers preparation for the app

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that runs `prepareHeaders` that configures the headers for React Native and Codegen targets.

## Changelog:
[Internal] -

Differential Revision: D81778440

Commit:
Add Function to update the PBXFrameworkBuildPhase

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change does two things:
- moves the printFilesForBuildPhase to xcode-core-utils to make it easier to mock the function
- Adds a function to update the package reference section to configure the list of build files for the FrameworkBuildStep

## Changelog:
[Internal] -

Differential Revision: D81778472

Commit:
Add Function to convert pbx file to JSON

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

Apple provides a helpful CLI program called `plutils` that can take a xcodeproj and convert it to json. We use that to parse the xcodeproject.

## Changelog:
[Internal] -

Differential Revision: D81778448

Commit:
Add Function to deintegrate local SwiftPM packages from Xcode project

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

We are adding a function that deintegrate local Swift Packages. This helps to make sure we can have an idempotent script and that we don't add the same package twice.

## Changelog:
[Internal] -

Differential Revision: D81778458

Commit:
Add helper functions to serialize specific objects of the project

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds some helper function that can be used to serialize some simple Xcodeproject objects back to string.
This will be used in future diffs to rewrite the Xcodeproj file.

## Changelog:
[Internal] -

Differential Revision: D81778445

Commit:
Add helper functions to add local packages to xcodeproject

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change does two things:
- moves the function that generate the object id to a different file
- adds a function to add the local Swift packages to the project

Moving the generateObjectId function to a separate file is a required step to mock it in the tests.

## Changelog:
[Internal] -

Differential Revision: D81934363

Commit:
Add script to find the directory that contains the Xcodeproj

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that given the root of the app and the name of the xcodeproject file, can return the path to the Xcode project file

## Changelog:
[Internal] -

Differential Revision: D81778456

Commit:
Add Function to update the package reference section

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds a function to update the package reference section so that Xcode can properly load the packages in the project.

## Changelog:
[Internal] -

Differential Revision: D81778463

Commit:
Add function to fix the REACT_NATIVE_PATH variable

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that fixes the REACT_NATIVE_PATH build settings. Cocoapods sets this value to something that starts with `PODS_ROOT`. However, without using Cocoapods, we can't really rely on that variable being set.

## Changelog:
[Internal] -

Differential Revision: D81778462

Commit:
Configure Package.swift file to be able to build from source (#53667)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53667

This change updates the Package.swift file to enable building RNTester and HelloWorld from source.

The set of changes is the following:
- add a BUILD_FROM_SOURCE flag to toggle between building for creating prebuilds and building an app from source
- configure the proper header search path based on whether we want to build an app from source or not
- split the ReactFabricComponents target in multiple subtargets. This is required because when building from source we can't have targets whose Source folder is overlapping.

## Changelog:
[Internal] -

Differential Revision: D81778452

Commit:
Add function to generate random Xcode ids

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

All the Xcode project elements have an id of 24 hex characters. This change adds a function to generate the id randomly.

## Changelog:
[Internal] -

Differential Revision: D81778473

Commit:
Add function to set build from source to true

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function updates the PAckage.swift file to set BUILD_FROM_SOURCE to true

## Changelog:
[Internal] -

Differential Revision: D81778460

Commit:
Add function to set the CLANG_ALLOW_NON_MODULAR_INCLUDES_IN_FRAMEWORK_MODULES build setting

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that set ALLOW_NON_MODULAR_HEADERS build settings to make the import work properly.

## Changelog:
[Internal] -

Differential Revision: D81778444

Commit:
Add Function to add missing sections in case the original project don't have them (#53648)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53648

## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds a function to add missing section to the project. This function is used by a coordination function in a future diff on the stack.

## Changelog:
[Internal] -

Differential Revision: D81778443

Commit:
Add function to configure the app for swift by creating modulemap and umbrella header

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that configure React Native to be swift compatible by creating the React-umbrella and modulemap

## Changelog:
[Internal] -

Differential Revision: D81778437

Commit:
Add Functions to coordinate the update of the xcode project.

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This is the last function required to manipulate the Xcodeproject file. It is an high-level function that orchestrate all the changes that are needed in the xcodeproj file

## Changelog:
[Internal] -

Differential Revision: D82020721

Commit:
Add Functions to manipulate the update of the xcode project.

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change:
- moves the print functions and the addMissingSections  function to a separate file, to make it easier to mocke them in tests
- add a function that coordinates the manipulation of the xcodeproj

## Changelog:
[Internal] -

Differential Revision: D81778451

Commit:
Add function to run pod deintegrate

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that runs `pod deintegrate` to remove remainings of cocoapods

## Changelog:
[Internal] -

Differential Revision: D81778468

Commit:
Add function to create hard links for the app

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that create hardlinks for React Native so it can build from source

## Changelog:
[Internal] -

Differential Revision: D81778454

Commit:
Add configure iOS prebuilds (hermes and dependencies)

Summary:
## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that prepares the prebuilds for ios so we can leverage them when building from source

## Changelog:
[Internal] -

Differential Revision: D81778467

cipolleschi committed:
Add function to open Xcodeproj (#53682)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53682

## Context

When configuring an app to build with SwiftPM from source, there is a sequence of operations we need to run in order to prepare the project correctly.

## Changed

Add a function that runs openx the xcodeproject once the integration is done.

## Changelog:
[Internal] -

Reviewed By: cortinico

Differential Revision: D81778466
